pkg/controller/storagenode/storagenode_test.go

```diff
@@ -0,0 +1,615 @@
+				UsedSize:  *resource.NewQuantity(10971520, resource.BinarySI),
+			},
+			Conditions: []corev1alpha1.NodeCondition{
+				{
+					Type:   corev1alpha1.NodeStateCondition,
+					Status: corev1alpha1.NodeOnlineStatus,
+				},
+			},
+		},
+	}
+
+	storageLessNode := &corev1alpha1.StorageNode{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            testStoragelessNode,
+			Namespace:       cluster.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*clusterRef},
+		},
+		Status: corev1alpha1.NodeStatus{
+			Phase: string(corev1alpha1.NodeInitStatus),
+			Storage: corev1alpha1.StorageStatus{
+				TotalSize: defaultQuantity,
+				UsedSize:  defaultQuantity,
+			},
+			Conditions: []corev1alpha1.NodeCondition{
+				{
+					Type:   corev1alpha1.NodeStateCondition,
+					Status: corev1alpha1.NodeOnlineStatus,
+				},
+			},
+		},
+	}
+
+	k8sClient := testutil.FakeK8sClient(storageNode, storageLessNode, cluster)
+	recorder := record.NewFakeRecorder(10)
+	driver := testutil.MockDriver(mockCtrl)
+	driver.EXPECT().GetSelectorLabels().Return(nil).AnyTimes()
+	driver.EXPECT().String().Return("ut-driver").AnyTimes()
+
+	// reconcile storage node
+	podNode1 := createStoragePod(cluster, "pod-node1", testStorageNode, nil, clusterRef)
+	k8sClient.Create(context.TODO(), podNode1)
+
+	podNode2 := createStoragePod(cluster, "pod-node2", testStoragelessNode, nil, clusterRef)
+	k8sClient.Create(context.TODO(), podNode2)
+
+	controller := Controller{
+		client:   k8sClient,
+		recorder: recorder,
+		Driver:   driver,
+	}
+
+	request := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      testStorageNode,
+			Namespace: testNS,
+		},
+	}
+	result, err := controller.Reconcile(request)
+	require.NoError(t, err)
+	require.NotNil(t, result)
+
+	checkStoragePod := &v1.Pod{}
+	err = controller.client.Get(context.TODO(), client.ObjectKey{
+		Name:      podNode1.Name,
+		Namespace: podNode1.Namespace,
+	}, checkStoragePod)
+	require.NoError(t, err)
+	value, present := checkStoragePod.Labels[constants.LabelKeyStoragePod]
+	require.True(t, present)
+	require.Equal(t, "true", value)
+
+	// now makes this node non-storage and ensure reconcile removes the label
+	storageNode.Status.Conditions = nil
+	err = controller.client.Update(context.TODO(), storageNode)
+	require.NoError(t, err)
+
+	result, err = controller.Reconcile(request)
+	require.NoError(t, err)
+	require.NotNil(t, result)
+
+	checkStoragePod = &v1.Pod{}
+	err = controller.client.Get(context.TODO(), client.ObjectKey{
+		Name:      podNode1.Name,
+		Namespace: podNode1.Namespace,
+	}, checkStoragePod)
+	require.NoError(t, err)
+	_, present = checkStoragePod.Labels[constants.LabelKeyStoragePod]
+	require.False(t, present)
+
+	// reconcile storageless node
+	request = reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      testStoragelessNode,
+			Namespace: testNS,
+		},
+	}
+	result, err = controller.Reconcile(request)
+	require.NoError(t, err)
+	require.NotNil(t, result)
```

Comment:
Shouldn't we check that the storagelessPod (podNode2)still does not have the labels?

Reply:
added check. It was already being checked indirectly above due to transition of storage node to storageless node.